### PR TITLE
[SPARK-37903][PYTHON][FOLLOW-UP] Raise ValueError with no return function

### PR DESCRIPTION
**Pull request by dchvn**
### What changes were proposed in this pull request?
Raise ValueError with no return function

### Why are the changes needed?
This PR is a follow-up of https://github.com/apache/spark/pull/35200
Currently, with the function with no return, `infer_return_type` will return `ScalarType[DoubleType]` as default. We should raise exception for this case.

### Does this PR introduce _any_ user-facing change?
Yes, 
Before this PR
```python
>>> from pyspark.pandas.typedef.typehints import infer_return_type
>>> def f():
...     pass
... 
>>> infer_return_type(f)
ScalarType[DoubleType]
```

after this PR user will take an exception when infer return type of function with no return.
```python-traceback
>>> from pyspark.pandas.typedef.typehints import infer_return_type
>>> def f():
...     pass
... 
>>> infer_return_type(f)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/u02/spark/python/pyspark/pandas/typedef/typehints.py", line 563, in infer_return_type
    raise ValueError("A return value is required for the input function")
ValueError: A return value is required for the input function

```

### How was this patch tested?
unit test

### CODE REVIEW

**Comment by dchvn** [comment]
CC @ueshin @HyukjinKwon FYI, Thanks!

File: python/pyspark/pandas/tests/test_typedef.py
Comment (itholic) [inline python/pyspark/pandas/tests/test_typedef.py:66]: Can we refine the message bit more explicitly ??

e.g. "A return value is required for the function passed to `infer_return_type`."

or something.

File: python/pyspark/pandas/tests/test_typedef.py
Comment (itholic) [inline python/pyspark/pandas/tests/test_typedef.py:74]: This too, It would be great if we can show the message more explicitly.

e.g. "None type is not acceptable for the function passed to `infer_return_type`."

or something.

**Comment by itholic** [comment]
Also add some more context to the PR description why this change is needed ??

At least It is good to provide a link which PR or comment you are following up.

File: python/pyspark/pandas/tests/test_typedef.py
Comment (dchvn) [reply python/pyspark/pandas/tests/test_typedef.py]: updated. Thanks!

File: python/pyspark/pandas/typedef/typehints.py
Comment (Yikun) [inline python/pyspark/pandas/typedef/typehints.py:563]: Technically, I think it should be ValueError, because func is `Callable` but the value is not not meet the requirement. See also in why raise TypeError before (it just because some compatiable reason):

https://github.com/python/cpython/blob/7c0914d35eaaab2f323260ba5fe8884732533888/Lib/inspect.py#L1374-L1379

It's not a strong prefer for me, just a reminder, you could do a double check on this, if you have other idea on this, I'm ok with this.

**Review by Yikun — APPROVED** [review]
LGTM, if you confirmed it's not a problem.

File: python/pyspark/pandas/typedef/typehints.py
Comment (dchvn) [reply python/pyspark/pandas/typedef/typehints.py]: That make sense! thanks!

**Comment by Yikun** [comment]
Thanks, remember to also update the PR description.

**Comment by HyukjinKwon** [comment]
Looks fine to me too but would be great to have a sign-off from @ueshin

**Review by itholic — APPROVED** [review]

**Review by ueshin — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by dchvn** [comment]
thanks all

**Comment by bjornjorgensen** [comment]
@dchvn I get some errors now when I use lamda to rename columns [SPARK-38763](https://issues.apache.org/jira/projects/SPARK/issues/SPARK-38763)

If I remove this PR then it works. 

I can`t see why we need this PR. 


@HyukjinKwon

**Comment by xinrong-databricks** [comment]
@bjornjorgensen I filed a fix against the issue you mentioned. Let's discuss further on the ticket if you have any questions :)

**Pull request closed without merging**
